File: packages/py-blockwise/src/blockwise/simulate.py
"""Simulate a blockwise missing pattern on complete data."""
from __future__ import annotations

from typing import Sequence, Union

import numpy as np
import pandas as pd

PropMissing = Union[float, Sequence[float]]


def simulate_blockwise_missing(
    data: pd.DataFrame,
    blocks: Sequence[Sequence[str]],
    prop_missing: PropMissing,
    noise: float = 0.05,
    seed: int = 1234,
) -> pd.DataFrame:
    """Joint-mask column groups on random subsets of rows.

    Parameters
    ----------
    data : DataFrame
        The complete dataset to mask.
    blocks : sequence of sequences of str
        Each element is a set of column names to mask *jointly* on one
        independent random sample of rows.
    prop_missing : float or sequence of floats
        Row fraction masked per block. Scalar (applied to each block) or
        length-``len(blocks)`` vector.
    noise : float, default=0.05
        Extra per-column random-NA rate, applied only to columns named
        in any block. Set to 0 to disable.
    seed : int, default=1234
        Base seed for reproducibility.

    Returns
    -------
    DataFrame
        Copy of ``data`` with ``NaN`` introduced in the specified pattern.
    """
    if not isinstance(data, pd.DataFrame):
        raise TypeError("`data` must be a pandas DataFrame.")
    if not isinstance(blocks, (list, tuple)) or len(blocks) == 0:
        raise ValueError("`blocks` must be a non-empty sequence of column lists.")

    if np.isscalar(prop_missing):
        prop_missing = [float(prop_missing)] * len(blocks)
    if len(prop_missing) != len(blocks):
        raise ValueError("`prop_missing` must be scalar or length-`len(blocks)`.")

    out = data.copy()
    n = len(out)

    for i, cols in enumerate(blocks):
        cols_i = [c for c in cols if c in out.columns]
        if not cols_i:
            continue
        rng = np.random.default_rng(seed + i)
        k = int(prop_missing[i] * n)
        if k <= 0:
            continue
        rows = rng.choice(n, size=k, replace=False)
        out.loc[out.index[rows], cols_i] = np.nan

    if noise > 0:
        affected = sorted({c for b in blocks for c in b if c in out.columns})
        for k_idx, col in enumerate(affected):
            rng = np.random.default_rng(seed + len(blocks) + k_idx)
            k = int(noise * n)
            if k <= 0:
                continue
            rows = rng.choice(n, size=k, replace=False)
            out.loc[out.index[rows], col] = np.nan

    return out

File: packages/r-blockwise/R/simulate.R
#' Simulate a blockwise missing pattern on otherwise complete data
#'
#' Joint-masks groups of columns on randomly chosen rows, optionally adding
#' light column-wise random-NA noise on top. Useful for benchmarking BRM
#' on complete datasets; the default arguments reproduce the simulation
#' design used in Srinivasan, Currim, and Ram (2025).
#'
#' @param data A data.frame.
#' @param blocks A list of character vectors; each vector names the columns
#'   masked jointly for one block of rows.
#' @param prop_missing Proportion of rows affected per block. Either a scalar
#'   (applied to each block) or a numeric vector of length
#'   \code{length(blocks)}.
#' @param noise Extra per-column random-NA rate applied on top, restricted to
#'   columns named in any block. Default 0.05. Set to 0 to disable.
#' @param seed Optional integer base seed. Default 1234.
#' @return A data.frame of the same shape as \code{data}, with \code{NA}s
#'   introduced in the specified pattern.
#' @examples
#' df <- data.frame(a = 1:100, b = 1:100, c = 1:100, d = 1:100)
#' simulate_blockwise_missing(df,
#'   blocks       = list(c("a", "b"), c("c", "d")),
#'   prop_missing = 0.3)
#' @export
simulate_blockwise_missing <- function(data, blocks,
                                       prop_missing,
                                       noise = 0.05,
                                       seed = 1234L) {
  stopifnot(is.data.frame(data), is.list(blocks), length(blocks) >= 1L)
  if (length(prop_missing) == 1L) {
    prop_missing <- rep(prop_missing, length(blocks))
  }
  if (length(prop_missing) != length(blocks)) {
    stop("`prop_missing` must be a scalar or length-`length(blocks)`.")
  }

  out <- data
  n <- nrow(data)

  for (i in seq_along(blocks)) {
    cols_i <- intersect(blocks[[i]], names(out))
    if (!length(cols_i)) next
    set.seed(seed + i)
    rows_i <- sample(n, floor(prop_missing[i] * n), replace = FALSE)
    out[rows_i, cols_i] <- NA
  }

  if (noise > 0) {
    affected <- unique(unlist(blocks))
    affected <- intersect(affected, names(out))
    for (k in seq_along(affected)) {
      set.seed(seed + length(blocks) + k)
      rows_k <- sample(n, floor(noise * n), replace = FALSE)
      out[rows_k, affected[k]] <- NA
    }
  }
  out
}
